src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -632,18 +631,19 @@ internal bool TryRead(out ReadResult result)
             }
         }
 
-        private static void TrySchedule(PipeScheduler scheduler, Action<object> action, object state)
+        private static void ScheduleCallbacks(PipeScheduler scheduler, PipeCompletionCallbacks completionCallbacks)
         {
-            if (action != null)
-            {
-                scheduler.UnsafeSchedule(action, state);
-            }
+            Debug.Assert(completionCallbacks != null);
+
+            scheduler.UnsafeSchedule(s_invokeCompletionCallbacks, completionCallbacks);
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Does it actually help with measurable throughput? You can make many methods disappear from call counts by marking them for aggressive inlining, but that doesn't necessarily mean it improves actual performance. Does this reduce the overall amount of generated native code?

Reply:
The checked-in GDV heuristics are pretty simple, so not it is not that surprising if they make poor guesses.

Reply:
Not sure guarded devirtualization uses the actual type of the instance at this stage, rather the declared type.
```
impDevirtualizeCall: Trying to devirtualize virtual call:
    class for 'this' is PipeScheduler (attrib 20000400)
    base method is PipeScheduler::UnsafeSchedule
    devirt to PipeScheduler::UnsafeSchedule -- inexact or not final
               [000700] --C-G-------              *  CALLV ind void   PipeScheduler.UnsafeSchedule
               [000693] ------------ this in rcx  +--*  LCL_VAR   ref    V33 tmp19        
               [000694] ------------ arg1         +--*  LCL_VAR   ref    V32 tmp18        
               [000699] --C--------- arg2         \--*  RET_EXPR  ref   (inl return from call [000697])
    Class not final or exact, and method not final
Marking call [000700] as guarded devirtualization candidate; will guess for class PipeScheduler
```

Reply:
```
Importing BB76 (PC=041) of 'Pipe:TrySchedule(ref,byref)'
    [ 0]  41 (0x029) ldarg.0
lvaGrabTemp returning 33 (V33 tmp19) called for Inlining Arg.
Marked V33 as a single def temp
Querying runtime about current class of field Pipe._writerScheduler (declared as PipeScheduler)
Field's current class not available

lvaSetClass: setting class for V33 to (00007FF9C58EB0D0) PipeScheduler 
```

Its coming from a field declared as the parent type
```
private readonly PipeScheduler _writerScheduler;
```
Which is then passed as a parameter
```
TrySchedule(_writerScheduler, completionData);
```
However its set in the .ctor via parameter, so has a more specific type at runtime, but not an observable complie time type. 

Not sure if its in the plan for guarded devirtualization to inspect the current instance's runtime type and use that for the compilation guess rather than declared type? (vs downside that first call might not be representative of the majority of following calls)

A way round it for this code could be to take a similar approach to the Dictionary devirtualization; where if it the value passed to the .ctor is a known type (here `ThreadPoolScheduler`); keep the field null. Then if the field is `null` make the more explicit call. e.g. "devritalize" via code but isn't very clean and only works for the type declared as part of the assembly (which is only one side)

Reply:
We have a bunch of things in mind for heuristics and looking at actual types at runtime is one of the things to consider.

As you say, not clear how effective it will be...